lib/awful/screenshot.lua

```diff
@@ -0,0 +1,881 @@
+-- @tparam[opt] string fp The full path to the filepath
+function screenshot:set_filepath(fp)
+  self:filepath_builder({filepath = fp})
+end
+
+--- Get screenshot method name
+--
+-- @property method_name
+function screenshot:get_method_name()
+  return self._private.method_name
+end
+
+--- Get screenshot screen
+--
+-- @property screen
+function screenshot:get_screen()
+  if self.method_name == "screen" then
+    return self._private.screen
+  else
+    return nil
+  end
+end
+
+--- Get screenshot client
+--
+-- @property client
+function screenshot:get_client()
+  if self.method_name == "client" then
+    return self._private.client
+  else
+    return nil
+  end
+end
+
+--- Get screenshot geometry
+--
+-- @property geometry
+function screenshot:get_geometry()
+  return self._private.geometry
+end
+
+--- Get screenshot surface
+--
+-- @property surface
+function screenshot:get_surface()
+  return self._private.surface
+end
+
+-- Methods for the screenshot object returned from taking a screenshot.
+
+--- Set the filepath to save a screenshot
+--
+-- @function awful.screenshot:filepath_builder
+-- @tparam[opt] table args Table with the filepath parameters
+function screenshot:filepath_builder(args)
+
+  local args = (type(args) == "table" and args) or {}
+
+  local filepath = args.filepath
+  local directory = args.directory
+  local prefix = args.prefix
+
+
+  if filepath and check_filepath(filepath) then
+
+    directory, prefix = parse_filepath(filepath)
+
+  elseif directory or prefix then
+
+    if directory and type(directory) == "string" then
+      directory = check_directory(directory)
+    elseif self.directory then
+      directory = self._private.directory -- The setter ran check_directory()
+    else
+      directory = get_default_dir()
+    end
+
+    if prefix and type(prefix) == "string" then
+      prefix = check_prefix(prefix)
+    elseif self.prefix then
+      prefix = self._private.prefix -- The setter ran check_prefix()
+    else
+      prefix = get_default_prefix()
+    end
+
+    directory, prefix, filepath = make_filepath(directory, prefix) 
+
+  elseif self.filepath and check_filepath(self.filepath) then
+
+    filepath = self.filepath
+    directory, prefix = parse_filepath(filepath)
+
+  else
+
+    if self.directory then
+      directory = self._private.directory -- The setter ran check_directory()
+    else
+      directory = get_default_dir()
+    end
+
+    if self.prefix then
+      prefix = self._private.prefix -- The setter ran check_prefix()
+    else
+      prefix = get_default_prefix()
+    end
+
+    directory, prefix, filepath = make_filepath(directory, prefix) 
+
+  end
+
+  if filepath then
+    self._private.directory = directory -- These have already
+    self._private.prefix = prefix       -- been checked.
+    self._private.filepath = filepath
+  end
+
+end
+
+--- Save screenshot
+--
+-- @function awful.screenshot:save
+-- @tparam[opt] table args Table with the filepath parameters
```

Comment:
I can do it that way too. It looked like you like to avoid having to deal with positional arguments and prefer the table instead. Remember that my path_builder takes both a directory and a prefix for the filename. Since I am working with table args, I also built in an explicitly passed full file path, which will be checked for validity (points to a valid writeable directory). The way I have it now, if the filepath argument is used, it will be used as is and does not append a date_time string.